src/base/bittorrent/private/speedmonitor.cpp

```diff
@@ -27,16 +27,24 @@
  * exception statement from your version.
  */
 
-#include <QList>
 #include "speedmonitor.h"
 
 void SpeedMonitor::addSample(const SpeedSample &sample)
 {
-    m_speedSamples.push_back(sample);
+    static_assert(MAX_SAMPLES > 0, "MAX_SAMPLES should > 0");
+
     m_sum += sample;
-    if (m_speedSamples.size() > MAX_SAMPLES) {
-        m_sum -= m_speedSamples.front();
-        m_speedSamples.pop_front();
+    if (m_speedSamples.size() >= MAX_SAMPLES) {
+        m_sum -= *m_samplesLast;
+        *m_samplesLast = sample;
+
+        ++m_samplesLast;
+        if (m_samplesLast == m_speedSamples.end())
+            m_samplesLast = m_speedSamples.begin();
+    }
+    else {
+        m_speedSamples.emplace_back(sample);
+        m_samplesLast = m_speedSamples.begin();
```

Comment:
What about initialize `m_samplesLast = m_speedSamples.end()` in constructor and then:
```c++
 if (m_speedSamples.size() >= MAX_SAMPLES) {
    if (m_samplesLast == m_speedSamples.end())
        m_samplesLast = m_speedSamples.begin();

    m_sum -= *m_samplesLast;
    *m_samplesLast = sample;
    ++m_samplesLast;
}
else {
    m_speedSamples.emplace_back(sample);
}
```